javalib/src/main/scala/java/lang/Byte.scala

```diff
@@ -225,12 +225,27 @@ object Byte {
   @inline def toUnsignedLong(x: scala.Byte): scala.Long =
     byteToULong(x)
 
-  @inline def valueOf(byteValue: scala.Byte): Byte =
-    new Byte(byteValue)
+  import ByteCache.cache
+
+  @inline def valueOf(byteValue: scala.Byte): Byte = {
+    val idx    = byteValue - MIN_VALUE
+    val cached = cache(idx)
+    if (cached != null) {
+      cached
+    } else {
+      val newbyte = new Byte(byteValue)
+      cache(idx) = newbyte
+      newbyte
+    }
+  }
 
   @inline def valueOf(s: String): Byte =
     valueOf(parseByte(s))
 
   @inline def valueOf(s: String, radix: scala.Int): Byte =
     valueOf(parseByte(s, radix))
 }
+
+private[lang] object ByteCache {
+  private[lang] val cache = new Array[java.lang.Byte](256)
```

Comment:
I suggest you fill in that cache immediately, so that you avoid the branching in `valueOf` altogether.

Same for the other caches, except of course for `Short`, `Int` and `Long` you still have the branch for whether it's small enough.